### PR TITLE
fix arrayjoin with some pipelines

### PR DESCRIPTION
**Pull request by jcupitt**
If the operation before arrayjoin was something that just forwarded a pointer, we could be left with dangling memory references in some cases.

Also some tiny comment adjustment and formatting changes.

thanks TheEssem

see https://github.com/libvips/libvips/issues/3835

### CODE REVIEW

**Comment by jcupitt** [comment]
Test program:

```python
#!/usr/bin/python3

import sys
import pyvips

original = pyvips.Image.new_from_file(sys.argv[1])
page_height = original.get_page_height()
n_pages = original.get_n_pages()

img = []
for i in range(0, n_pages):
  x = i * page_height
  cropped = original.crop(0, x, original.width, page_height // 2)
  flipped = cropped.flip('vertical')
  img.append(cropped)
  img.append(flipped)

image = pyvips.Image.arrayjoin(img, across=1)
image.set('page-height', 2 * (page_height // 2))
image.write_to_file(sys.argv[2])
```

Run with eg.:

```
$ ./mirror.py ~/pics/3198.gif[n=-1] x.gif
```

With larger GIFs (over 100mb fully decompressed) you'll get a segv without this PR.

File: libvips/conversion/arrayjoin.c [hunk 1]
```diff
@@ -112,16 +114,24 @@ vips_arrayjoin_gen(VipsRegion *out_region,
...
 		i = VIPS_MIN(n - 1, left + top * join->across);
 
-		reg = vips_region_new(in[i]);
+		/* The part of in[i] we need.
+		 */
+		need = out_region->valid;
+		need.left -= join->rects[i].left;
+		need.top -= join->rects[i].top;
 
-		if (vips__insert_just_one(out_region, reg,
```
Comment (kleisauke) [inline libvips/conversion/arrayjoin.c:119]: It looks like `vips__insert_just_one()` can be made static after this.
https://github.com/libvips/libvips/blob/3698b934969b00e8d7497186d64f8f18a49b68cf/libvips/conversion/insert.c#L118-L123

**Review by kleisauke — APPROVED** [review]
LGTM! I left a minor comment inline, but that can also be addressed after this PR lands.

File: libvips/conversion/arrayjoin.c (hunk 1 above)
Comment (jcupitt) [reply libvips/conversion/arrayjoin.c:119]: Ah good point, I'll do that too.

**Pull request merged**